src/mlpack/methods/ann/augmented/ham_unit_impl.hpp

```diff
@@ -0,0 +1,99 @@
+/**
+ * @file ham_unit_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of HAMUnit class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_AUGMENTED_HAM_UNIT_IMPL_HPP
+#define MLPACK_METHODS_AUGMENTED_HAM_UNIT_IMPL_HPP
+
+#include "ham_unit.hpp"
+
+namespace mlpack {
+namespace ann /* Artificial Neural Network */ {
+namespace augmented /* Augmented neural network */ {
+
+HAMUnit::HAMUnit(int memorySize,
+                 C& controller,
+                 E& embed,
+                 J& join,
+                 S& search,
+                 W& write)
+  : memorySize(memorySize),
+    search(search), embed(embed), controller(controller), t(0)
+{
+  memory = TreeMemory<double, J, W>(memorySize, join, write);
+}
+
+arma::vec HAMUnit::Attention() const
+{
+  size_t nodesCnt = 2 * memory.ActualMemorySize() - 1;
+  arma::vec probabilities(nodesCnt);
+  probabilities(0) = 1;
+  arma::vec hController = sequence.col(t++);
+  for (size_t node = 1; node < nodesCnt; ++node) {
+    size_t parent = memory.Parent(node);
+    bool dir = node == memory.Left(parent);
+    arma::vec h(hController.n_elem + memory.GetCell(parent).n_elem);
+    h.rows(0, hController.n_elem - 1) = hController;
+    h.rows(hController.n_elem,
+           hController.n_elem + memory.GetCell(parent).n_elem - 1)
+        = memory.GetCell(parent);
+    double prob = search.Forward(h);
+    if (!dir) prob = 1. - prob;
+    probabilities(node) = prob * probabilities(parent);
```

Comment:
Does this implement Figure 3 from the HAM paper correctly?  I thought that at a given node in the tree, we calculated p = SEARCH(h_node, h_lstm), and then went left with probability (1 - p) and went right with probability p.  So we would have to same a random number to do that.  Then when we reach a leaf, the value held in the TreeMemory there would be what we use as the input for the LSTM for the output stage.  Maybe I have read this wrong, but it seems like this does not implement that.  Or is this implementing 'soft attention' from the appendix?